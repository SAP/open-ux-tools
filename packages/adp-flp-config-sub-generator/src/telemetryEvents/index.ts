/**
 * Event names for telemetry for the adaptation project fiori launchpad configuration generator
 */
export enum EventName {
    ADP_FLP_CONFIG_ADDED = 'ADP_FLP_CONFIG_ADDED'
}
