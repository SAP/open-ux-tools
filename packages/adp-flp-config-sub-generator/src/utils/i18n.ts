import type { TOptions } from 'i18next';
import i18next from 'i18next';
import translations from '../translations/adp-flp-config-sub-generator.i18n.json';

const adpFlpConfigI18nNamespace = 'adp-flp-config-sub-generator';

/**
 * Initialize i18next with the translations for this module.
 */
export async function initI18n(): Promise<void> {
    await i18next.init({ lng: 'en', fallbackLng: 'en' }, () =>
        i18next.addResourceBundle('en', adpFlpConfigI18nNamespace, translations)
    );
}

/**
 * Helper function facading the call to i18next. Unless a namespace option is provided the local namespace will be used.
 *
 * @param key i18n key
 * @param options additional options
 * @returns {string} localized string stored for the given key
 */
export function t(key: string, options?: TOptions): string {
    if (!options?.ns) {
        options = Object.assign(options ?? {}, { ns: adpFlpConfigI18nNamespace });
    }
    return i18next.t(key, options);
}

initI18n().catch(() => {
    // Needed for lint
});
