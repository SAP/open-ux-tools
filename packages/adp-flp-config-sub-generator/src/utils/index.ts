export * from './provider';
export * from './logger';
export * from './appWizardCache';
export * from './i18n';
