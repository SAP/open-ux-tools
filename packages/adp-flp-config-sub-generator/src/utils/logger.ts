import { DefaultLogger, LogWrapper, type ILogWrapper } from '@sap-ux/fiori-generator-shared';
import type { Logger } from 'yeoman-environment';
import type { IVSCodeExtLogger, LogLevel } from '@vscode-logging/logger';

/**
 * Static logger prevents passing of logger references through all functions, as this is a cross-cutting concern.
 */
export default class AdpFlpConfigLogger {
    private static _logger: ILogWrapper = DefaultLogger;

    /**
     * Get the logger.
     *
     * @returns the logger
     */
    public static get logger(): ILogWrapper {
        return AdpFlpConfigLogger._logger;
    }

    /**
     * Set the logger.
     *
     * @param value the logger to set
     */
    public static set logger(value: ILogWrapper) {
        AdpFlpConfigLogger._logger = value;
    }

    /**
     * Configures the vscode logger.
     *
     * @param vscLogger - the vscode logger
     * @param loggerName - the logger name
     * @param yoLogger - the yeoman logger
     * @param vscode - the vscode instance
     * @param logLevel - the log level
     */
    static configureLogging(
        vscLogger: IVSCodeExtLogger,
        loggerName: string,
        yoLogger: Logger,
        vscode?: unknown,
        logLevel?: LogLevel
    ): void {
        const logWrapper = new LogWrapper(loggerName, yoLogger, logLevel, vscLogger, vscode);
        AdpFlpConfigLogger.logger = logWrapper;
    }
}
