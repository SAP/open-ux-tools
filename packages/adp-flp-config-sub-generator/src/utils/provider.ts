import { type AxiosRequestConfig, type ProviderConfiguration, type AbapServiceProvider } from '@sap-ux/axios-extension';
import { type AdpPreviewConfig } from '@sap-ux/adp-tooling';
import { type CredentialsAnswers } from '@sap-ux/inquirer-common';
import type { ToolsLogger } from '@sap-ux/logger';
import { createAbapServiceProvider } from '@sap-ux/system-access';

/**
 * Creates and returns an instance of AbapServiceProvider using the current UI5 YAML configuration and credentials.
 *
 * @param {AdpPreviewConfig} ui5Yaml - The UI5 YAML configuration containing the target ABAP system details.
 * @param {ToolsLogger} logger - The logger instance for logging messages.
 * @param {CredentialsAnswers} [credentials] - Optional credentials for authentication against the ABAP system.
 * @returns {Promise<AbapServiceProvider>} The ABAP service provider instance.
 */
export async function getAbapServiceProvider(
    ui5Yaml: AdpPreviewConfig,
    logger: ToolsLogger,
    credentials?: CredentialsAnswers
): Promise<AbapServiceProvider> {
    const { target, ignoreCertErrors = false } = ui5Yaml;
    const requestOptions: AxiosRequestConfig & Partial<ProviderConfiguration> = { ignoreCertErrors };
    if (credentials) {
        requestOptions['auth'] = { username: credentials.username, password: credentials.password };
    }
    return createAbapServiceProvider(target, requestOptions, false, logger);
}
