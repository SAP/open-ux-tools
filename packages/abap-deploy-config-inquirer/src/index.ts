import { ToolsLogger, type Logger } from '@sap-ux/logger';
import { initI18n } from './i18n';
import { PromptState } from './prompts/prompt-state';
import { getAbapDeployConfigQuestions } from './prompts';
import LoggerHelper from './logger-helper';
import type { InquirerAdapter } from '@sap-ux/inquirer-common';
import type { AbapDeployConfigAnswers, AbapDeployConfigPromptOptions } from './types';

/**
 * Get the inquirer prompts for abap deploy config.
 *
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompts used to provide input for abap deploy config generation and a reference to the answers object which will be populated with the user's responses once `inquirer.prompt` returns
 */
async function getPrompts(
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<{ prompts: any[]; answers: Partial<AbapDeployConfigAnswers> }> {
    await initI18n();
    LoggerHelper.logger = logger ?? new ToolsLogger({ logPrefix: '@sap-ux/abap-deploy-config-inquirer' });

    return {
        prompts: await getAbapDeployConfigQuestions(promptOptions),
        // Return reference to derived answers object that will be populated with user responses (after prompting is complete)
        answers: PromptState.abapDeployConfig
    };
}

/**
 * Prompt for abap deploy config writer inputs.
 *
 * @param adapter - optionally provide references to a calling inquirer instance, this supports integration to Yeoman generators, for example
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompt answers
 */
async function prompt(
    adapter: InquirerAdapter,
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<AbapDeployConfigAnswers> {
    const abapDeployConfigPrompts = (await getPrompts(promptOptions, logger)).prompts;
    const answers = await adapter.prompt<AbapDeployConfigAnswers>(abapDeployConfigPrompts);

    // Add dervied service answers to the answers object
    Object.assign(answers, PromptState.abapDeployConfig);

    return answers;
}

export { getPrompts, prompt, type InquirerAdapter, type AbapDeployConfigAnswers, type AbapDeployConfigPromptOptions };
