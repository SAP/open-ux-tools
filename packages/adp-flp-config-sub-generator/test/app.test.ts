import type { BackendSystem } from '@sap-ux/store';
import type { YUIQuestion, CredentialsAnswers } from '@sap-ux/inquirer-common';
import type { FLPConfigAnswers } from '@sap-ux/flp-config-inquirer';
import type { ToolsLogger } from '@sap-ux/logger';
import { join } from 'path';
import yeomanTest from 'yeoman-test';
import fs from 'fs';
import fsextra from 'fs-extra';
import adpFlpConfigGenerator from '../src/app';
import * as adpTooling from '@sap-ux/adp-tooling';
import * as btpUtils from '@sap-ux/btp-utils';
import * as Logger from '@sap-ux/logger';
import * as fioriGenShared from '@sap-ux/fiori-generator-shared';
import { rimraf } from 'rimraf';
import { EventName } from '../src/telemetryEvents';
import * as sysAccess from '@sap-ux/system-access';
import { t } from '../src/utils/i18n';
import { MessageType } from '@sap-devx/yeoman-ui-types';
import * as inquirerCommon from '@sap-ux/inquirer-common';
import * as projectAccess from '@sap-ux/project-access';

jest.mock('@sap-ux/system-access');
jest.mock('@sap-ux/btp-utils');
jest.mock('@sap-ux/adp-tooling', () => ({
    ...jest.requireActual('@sap-ux/adp-tooling'),
    isCFEnvironment: jest.fn(),
    getAdpConfig: jest.fn(),
    generateInboundConfig: jest.fn()
}));
jest.mock('@sap-ux/inquirer-common', () => ({
    ...jest.requireActual('@sap-ux/inquirer-common'),
    getCredentialsPrompts: jest.fn()
}));
jest.mock('@sap-ux/fiori-generator-shared', () => ({
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    ...(jest.requireActual('@sap-ux/fiori-generator-shared') as {}),
    sendTelemetry: jest.fn().mockReturnValue(new Promise(() => {})),
    TelemetryHelper: {
        initTelemetrySettings: jest.fn(),
        createTelemetryData: jest.fn().mockReturnValue({
            OperatingSystem: 'testOS',
            Platform: 'testPlatform'
        })
    },
    isExtensionInstalled: jest.fn().mockReturnValue(true),
    getHostEnvironment: jest.fn(),
    isCli: jest.fn().mockReturnValue(false)
}));

const loggerMock: ToolsLogger = {
    debug: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn()
} as Partial<ToolsLogger> as ToolsLogger;
jest.spyOn(Logger, 'ToolsLogger').mockImplementation(() => loggerMock);

describe('FLPConfigGenerator Integration Tests', () => {
    jest.setTimeout(100000);
    jest.spyOn(adpTooling, 'isCFEnvironment').mockReturnValue(false);

    const generatorPath = join(__dirname, '../../src/app/index.ts');
    const testOutputDir = join(__dirname, 'test-output');
    const credentialsPrompts = {
        prompts: [
            {
                username: 'systemUsername'
            },
            {
                password: 'systemPassword'
            }
        ]
    };
    let answers:
        | FLPConfigAnswers
        | {
              username: string;
              password: string;
          };

    jest.spyOn(adpTooling.ManifestService, 'initMergedManifest').mockResolvedValue({
        getManifest: jest.fn().mockReturnValue({})
    } as unknown as adpTooling.ManifestService);
    const showInformationSpy = jest.fn();
    const mockAppWizard = {
        setHeaderTitle: jest.fn(),
        showInformation: showInformationSpy
    };
    jest.spyOn(inquirerCommon, 'getCredentialsPrompts').mockResolvedValue(
        credentialsPrompts as unknown as YUIQuestion<CredentialsAnswers>[]
    );
    const vsCodeMessageSpy = jest.fn();
    const vscode = {
        window: {
            showErrorMessage: vsCodeMessageSpy
        }
    };
    jest.spyOn(projectAccess, 'getAppType').mockResolvedValue('Fiori Adaptation');

    beforeEach(() => {
        answers = {
            username: 'testUsername',
            password: 'testPassword',
            semanticObject: 'testSemanticObject',
            emptyInboundsInfo: 'testEmptyInboundsInfo',
            action: 'testAction',
            title: 'testTitle',
            subTitle: 'testSubTitle',
            additionalParameters: 'param1=test1&param2=test2'
        };
    });

    beforeAll(() => {
        fs.mkdirSync(testOutputDir, { recursive: true });
    });

    afterAll(() => {
        rimraf.sync(testOutputDir);
    });

    it('should generate FLP configuration successfully - Application Studio', async () => {
        const testPath = join(testOutputDir, 'test_project1');
        fs.mkdirSync(testPath, { recursive: true });
        fsextra.copySync(join(__dirname, 'fixtures/app.variant1'), join(testPath, 'app.variant1'));
        const testProjectPath = join(testPath, 'app.variant1');

        jest.spyOn(adpTooling, 'getAdpConfig').mockResolvedValue({
            target: {
                destination: 'testDestination'
            }
        });
        jest.spyOn(btpUtils, 'isAppStudio').mockReturnValue(true);
        jest.spyOn(btpUtils, 'listDestinations').mockResolvedValue({
            testDestination: {
                Name: 'https://testUrl',
                Host: '000',
                Type: 'Type',
                Authentication: 'Basic',
                ProxyType: 'Internet',
                Description: 'Description'
            }
        });
        const sendTelemetrySpy = jest.spyOn(fioriGenShared, 'sendTelemetry');

        const runContext = yeomanTest
            .create(
                adpFlpConfigGenerator,
                {
                    resolved: generatorPath
                },
                {
                    cwd: testProjectPath
                }
            )
            .withOptions({
                vscode,
                appWizard: mockAppWizard,
                launchAsSubGen: false
            })
            .withPrompts(answers);

        await expect(runContext.run()).resolves.not.toThrow();
        const variant = fs.readFileSync(join(testProjectPath, 'webapp', 'manifest.appdescr_variant'), {
            encoding: 'utf8'
        });
        const i18n = fs.readFileSync(join(testProjectPath, 'webapp', 'i18n', 'i18n.properties'), {
            encoding: 'utf8'
        });
        expect(variant).toMatchSnapshot();
        expect(i18n).toMatchSnapshot();
        expect(sendTelemetrySpy).toBeCalledWith(
            EventName.ADP_FLP_CONFIG_ADDED,
            expect.objectContaining({
                OperatingSystem: 'testOS',
                Platform: 'testPlatform'
            }),
            testProjectPath
        );
        expect(showInformationSpy).toHaveBeenCalledWith(t('info.flpConfigAdded'), MessageType.notification);
    });

    it('should generate FLP configuration successfully - VS Code', async () => {
        showInformationSpy.mockReset();
        const testPath = join(testOutputDir, 'test_project2');
        fs.mkdirSync(testPath, { recursive: true });
        fsextra.copySync(join(__dirname, 'fixtures/app.variant1'), join(testPath, 'app.variant1'));
        const testProjectPath = join(testPath, 'app.variant1');
        jest.spyOn(sysAccess, 'getCredentialsFromStore').mockResolvedValue({
            name: 'testSystem'
        } as unknown as BackendSystem);
        jest.spyOn(adpTooling, 'getAdpConfig').mockResolvedValue({
            target: {
                url: 'https://testUrl',
                client: '000'
            }
        });
        jest.spyOn(btpUtils, 'isAppStudio').mockReturnValue(false);
        const sendTelemetrySpy = jest.spyOn(fioriGenShared, 'sendTelemetry').mockRejectedValueOnce(new Error('Error'));

        const runContext = yeomanTest
            .create(
                adpFlpConfigGenerator,
                {
                    resolved: generatorPath
                },
                {
                    cwd: testProjectPath
                }
            )
            .withOptions({
                vscode,
                appWizard: mockAppWizard,
                launchAsSubGen: true
            })
            .withPrompts(answers);

        await expect(runContext.run()).resolves.not.toThrow();
        const variant = fs.readFileSync(join(testProjectPath, 'webapp', 'manifest.appdescr_variant'), {
            encoding: 'utf8'
        });
        const i18n = fs.readFileSync(join(testProjectPath, 'webapp', 'i18n', 'i18n.properties'), {
            encoding: 'utf8'
        });
        expect(variant).toMatchSnapshot();
        expect(i18n).toMatchSnapshot();
        expect(sendTelemetrySpy).toBeCalledWith(
            EventName.ADP_FLP_CONFIG_ADDED,
            expect.objectContaining({
                OperatingSystem: 'testOS',
                Platform: 'testPlatform'
            }),
            testProjectPath
        );
        expect(showInformationSpy).not.toBeCalled();
    });

    it('Should result in an error message if the project is a CF project', async () => {
        jest.spyOn(adpTooling, 'isCFEnvironment').mockReturnValueOnce(true);
        jest.spyOn(adpTooling, 'getAdpConfig').mockResolvedValue({
            target: {
                destination: 'testDestination'
            }
        });
        jest.spyOn(btpUtils, 'isAppStudio').mockReturnValue(true);
        const testProjectPath = join(__dirname, 'fixtures/app.variant1');

        const runContext = yeomanTest
            .create(
                adpFlpConfigGenerator,
                {
                    resolved: generatorPath
                },
                {
                    cwd: testProjectPath
                }
            )
            .withOptions({
                vscode,
                appWizard: mockAppWizard,
                launchFlpConfigAsSubGenerator: false
            })
            .withPrompts(answers);

        await expect(runContext.run()).rejects.toThrow(t('error.projectNotSupported'));
    });

    it('Should throw an error if writing phase fails', async () => {
        jest.spyOn(adpTooling, 'getAdpConfig').mockResolvedValue({
            target: {
                destination: 'testDestination'
            }
        });
        jest.spyOn(adpTooling, 'generateInboundConfig').mockRejectedValueOnce(new Error('Error'));
        jest.spyOn(btpUtils, 'isAppStudio').mockReturnValue(true);
        const testProjectPath = join(__dirname, 'fixtures/app.variant1');

        const runContext = yeomanTest
            .create(
                adpFlpConfigGenerator,
                {
                    resolved: generatorPath
                },
                {
                    cwd: testProjectPath
                }
            )
            .withOptions({
                vscode,
                appWizard: mockAppWizard,
                launchFlpConfigAsSubGenerator: false
            })
            .withPrompts(answers);

        await expect(runContext.run()).rejects.toThrow(t('error.updatingApp'));
    });

    it('Should throw an error when no destination is configured in Application Studio', async () => {
        jest.spyOn(adpTooling, 'getAdpConfig').mockResolvedValue({
            target: {} as unknown as sysAccess.AbapTarget
        });
        jest.spyOn(btpUtils, 'isAppStudio').mockReturnValue(true);
        jest.spyOn(btpUtils, 'listDestinations').mockResolvedValue({});
        const testProjectPath = join(__dirname, 'fixtures/app.variant1');

        const runContext = yeomanTest
            .create(
                adpFlpConfigGenerator,
                {
                    resolved: generatorPath
                },
                {
                    cwd: testProjectPath
                }
            )
            .withOptions({
                vscode,
                appWizard: mockAppWizard,
                launchFlpConfigAsSubGenerator: false
            })
            .withPrompts(answers);

        await runContext.run();
        expect(vsCodeMessageSpy).toBeCalledWith(t('error.destinationNotFound'));
    });

    it('Should throw an error when no url is configured for target in VS Code', async () => {
        jest.spyOn(adpTooling, 'getAdpConfig').mockResolvedValue({
            target: {} as unknown as sysAccess.AbapTarget
        });
        jest.spyOn(btpUtils, 'isAppStudio').mockReturnValue(false);
        const testProjectPath = join(__dirname, 'fixtures/app.variant1');

        const runContext = yeomanTest
            .create(
                adpFlpConfigGenerator,
                {
                    resolved: generatorPath
                },
                {
                    cwd: testProjectPath
                }
            )
            .withOptions({
                vscode,
                appWizard: mockAppWizard,
                launchFlpConfigAsSubGenerator: false
            })
            .withPrompts(answers);

        await runContext.run();
        expect(vsCodeMessageSpy).toBeCalledWith(t('error.systemNotFound'));
    });

    it('Should throw an error when system is not found in the store in VS Code', async () => {
        const systemUrl = 'https://testUrl';
        jest.spyOn(adpTooling, 'getAdpConfig').mockResolvedValue({
            target: {
                url: systemUrl
            }
        });
        jest.spyOn(sysAccess, 'getCredentialsFromStore').mockResolvedValueOnce(undefined);
        jest.spyOn(btpUtils, 'isAppStudio').mockReturnValue(false);
        const testProjectPath = join(__dirname, 'fixtures/app.variant1');

        const runContext = yeomanTest
            .create(
                adpFlpConfigGenerator,
                {
                    resolved: generatorPath
                },
                {
                    cwd: testProjectPath
                }
            )
            .withOptions({
                vscode,
                appWizard: mockAppWizard,
                launchFlpConfigAsSubGenerator: false
            })
            .withPrompts(answers);
        await runContext.run();
        expect(vsCodeMessageSpy).toBeCalledWith(t('error.systemNotFound', { systemUrl }));
    });

    it('Should throw an error when fetching manifest fails', async () => {
        jest.spyOn(adpTooling, 'getAdpConfig').mockResolvedValue({
            target: {
                url: 'https://testUrl'
            }
        });
        jest.spyOn(adpTooling.ManifestService, 'initMergedManifest').mockRejectedValueOnce(new Error('Error'));
        jest.spyOn(btpUtils, 'isAppStudio').mockReturnValue(false);
        const testProjectPath = join(__dirname, 'fixtures/app.variant1');

        const runContext = yeomanTest
            .create(
                adpFlpConfigGenerator,
                {
                    resolved: generatorPath
                },
                {
                    cwd: testProjectPath
                }
            )
            .withOptions({
                vscode,
                appWizard: mockAppWizard,
                launchFlpConfigAsSubGenerator: false
            })
            .withPrompts(answers);

        await expect(runContext.run()).rejects.toThrow(t('error.fetchingManifest'));
    });
});
