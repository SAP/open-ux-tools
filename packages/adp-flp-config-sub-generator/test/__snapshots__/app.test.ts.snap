// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FLPConfigGenerator Integration Tests should generate FLP configuration successfully - Application Studio 1`] = `
"{
  "fileName": "manifest",
  "layer": "CUSTOMER_BASE",
  "fileType": "appdescr_variant",
  "reference": "mockReference",
  "id": "customer.app.variant1",
  "namespace": "apps/mockReference/appVariants/customer.app.variant1/",
  "version": "0.1.0",
  "content": [
    {
      "changeType": "appdescr_ui5_addNewModelEnhanceWith",
      "content": {
        "modelId": "i18n"
      },
      "texts": {
        "i18n": "i18n/i18n.properties"
      }
    },
    {
      "changeType": "appdescr_ui5_addNewModelEnhanceWith",
      "content": {
        "modelId": "@i18n"
      },
      "texts": {
        "i18n": "i18n/i18n.properties"
      }
    },
    {
      "changeType": "appdescr_ui5_addNewModelEnhanceWith",
      "content": {
        "modelId": "mockModel"
      },
      "texts": {
        "i18n": "mockI18n"
      }
    },
    {
      "changeType": "appdescr_ui5_addNewModelEnhanceWith",
      "content": {
        "modelId": "mockModel"
      },
      "texts": {
        "i18n": "mockI18n"
      }
    },
    {
      "changeType": "appdescr_ui5_setMinUI5Version",
      "content": {
        "minUI5Version": "1.120.25"
      }
    },
    {
      "changeType": "appdescr_app_setTitle",
      "content": {},
      "texts": {
        "i18n": "i18n/i18n.properties"
      }
    },
    {
      "changeType": "appdescr_app_addNewInbound",
      "content": {
        "inbound": {
          "customer.app.variant1.InboundID": {
            "action": "testAction",
            "semanticObject": "testSemanticObject",
            "title": "{{customer.app.variant1_sap.app.crossNavigation.inbounds.customer.app.variant1.InboundID.title}}",
            "signature": {
              "additionalParameters": "allowed",
              "parameters": {
                "param1": {
                  "required": true,
                  "defaultValue": {
                    "value": "test1",
                    "format": "plain"
                  }
                },
                "param2": {
                  "required": true,
                  "defaultValue": {
                    "value": "test2",
                    "format": "plain"
                  }
                },
                "sap-appvar-id": {
                  "required": true,
                  "filter": {
                    "value": "customer.app.variant1",
                    "format": "plain"
                  },
                  "launcherValue": {
                    "value": "customer.app.variant1"
                  }
                }
              }
            },
            "subTitle": "{{customer.app.variant1_sap.app.crossNavigation.inbounds.customer.app.variant1.InboundID.subTitle}}"
          }
        }
      },
      "texts": {
        "i18n": "i18n/i18n.properties"
      }
    },
    {
      "changeType": "appdescr_app_removeAllInboundsExceptOne",
      "content": {
        "inboundId": "customer.app.variant1.InboundID"
      },
      "texts": {}
    },
    {
      "changeType": "appdescr_app_addNewInbound",
      "content": {
        "inbound": {
          "customer.app.variant1.InboundID": {
            "action": "testAction",
            "semanticObject": "testSemanticObject",
            "title": "{{customer.app.variant1_sap.app.crossNavigation.inbounds.customer.app.variant1.InboundID.title}}",
            "signature": {
              "additionalParameters": "allowed",
              "parameters": {
                "param1": {
                  "required": true,
                  "defaultValue": {
                    "value": "test1",
                    "format": "plain"
                  }
                },
                "param2": {
                  "required": true,
                  "defaultValue": {
                    "value": "test2",
                    "format": "plain"
                  }
                },
                "sap-appvar-id": {
                  "required": true,
                  "filter": {
                    "value": "customer.app.variant1",
                    "format": "plain"
                  },
                  "launcherValue": {
                    "value": "customer.app.variant1"
                  }
                }
              }
            },
            "subTitle": "{{customer.app.variant1_sap.app.crossNavigation.inbounds.customer.app.variant1.InboundID.subTitle}}"
          }
        }
      },
      "texts": {
        "i18n": "i18n/i18n.properties"
      }
    },
    {
      "changeType": "appdescr_app_removeAllInboundsExceptOne",
      "content": {
        "inboundId": "customer.app.variant1.InboundID"
      },
      "texts": {}
    }
  ]
}
"
`;

exports[`FLPConfigGenerator Integration Tests should generate FLP configuration successfully - Application Studio 2`] = `
"#Make sure you provide a unique prefix to the newly added keys in this file, to avoid overriding of SAP Fiori application keys.

#XTIT: Application name
customer.app.variant1_sap.app.title=App Variant Title"
`;

exports[`FLPConfigGenerator Integration Tests should generate FLP configuration successfully - VS Code 1`] = `
"{
  "fileName": "manifest",
  "layer": "CUSTOMER_BASE",
  "fileType": "appdescr_variant",
  "reference": "mockReference",
  "id": "customer.app.variant1",
  "namespace": "apps/mockReference/appVariants/customer.app.variant1/",
  "version": "0.1.0",
  "content": [
    {
      "changeType": "appdescr_ui5_addNewModelEnhanceWith",
      "content": {
        "modelId": "i18n"
      },
      "texts": {
        "i18n": "i18n/i18n.properties"
      }
    },
    {
      "changeType": "appdescr_ui5_addNewModelEnhanceWith",
      "content": {
        "modelId": "@i18n"
      },
      "texts": {
        "i18n": "i18n/i18n.properties"
      }
    },
    {
      "changeType": "appdescr_ui5_addNewModelEnhanceWith",
      "content": {
        "modelId": "mockModel"
      },
      "texts": {
        "i18n": "mockI18n"
      }
    },
    {
      "changeType": "appdescr_ui5_addNewModelEnhanceWith",
      "content": {
        "modelId": "mockModel"
      },
      "texts": {
        "i18n": "mockI18n"
      }
    },
    {
      "changeType": "appdescr_ui5_setMinUI5Version",
      "content": {
        "minUI5Version": "1.120.25"
      }
    },
    {
      "changeType": "appdescr_app_setTitle",
      "content": {},
      "texts": {
        "i18n": "i18n/i18n.properties"
      }
    },
    {
      "changeType": "appdescr_app_addNewInbound",
      "content": {
        "inbound": {
          "customer.app.variant1.InboundID": {
            "action": "testAction",
            "semanticObject": "testSemanticObject",
            "title": "{{customer.app.variant1_sap.app.crossNavigation.inbounds.customer.app.variant1.InboundID.title}}",
            "signature": {
              "additionalParameters": "allowed",
              "parameters": {
                "param1": {
                  "required": true,
                  "defaultValue": {
                    "value": "test1",
                    "format": "plain"
                  }
                },
                "param2": {
                  "required": true,
                  "defaultValue": {
                    "value": "test2",
                    "format": "plain"
                  }
                },
                "sap-appvar-id": {
                  "required": true,
                  "filter": {
                    "value": "customer.app.variant1",
                    "format": "plain"
                  },
                  "launcherValue": {
                    "value": "customer.app.variant1"
                  }
                }
              }
            },
            "subTitle": "{{customer.app.variant1_sap.app.crossNavigation.inbounds.customer.app.variant1.InboundID.subTitle}}"
          }
        }
      },
      "texts": {
        "i18n": "i18n/i18n.properties"
      }
    },
    {
      "changeType": "appdescr_app_removeAllInboundsExceptOne",
      "content": {
        "inboundId": "customer.app.variant1.InboundID"
      },
      "texts": {}
    },
    {
      "changeType": "appdescr_app_addNewInbound",
      "content": {
        "inbound": {
          "customer.app.variant1.InboundID": {
            "action": "testAction",
            "semanticObject": "testSemanticObject",
            "title": "{{customer.app.variant1_sap.app.crossNavigation.inbounds.customer.app.variant1.InboundID.title}}",
            "signature": {
              "additionalParameters": "allowed",
              "parameters": {
                "param1": {
                  "required": true,
                  "defaultValue": {
                    "value": "test1",
                    "format": "plain"
                  }
                },
                "param2": {
                  "required": true,
                  "defaultValue": {
                    "value": "test2",
                    "format": "plain"
                  }
                },
                "sap-appvar-id": {
                  "required": true,
                  "filter": {
                    "value": "customer.app.variant1",
                    "format": "plain"
                  },
                  "launcherValue": {
                    "value": "customer.app.variant1"
                  }
                }
              }
            },
            "subTitle": "{{customer.app.variant1_sap.app.crossNavigation.inbounds.customer.app.variant1.InboundID.subTitle}}"
          }
        }
      },
      "texts": {
        "i18n": "i18n/i18n.properties"
      }
    },
    {
      "changeType": "appdescr_app_removeAllInboundsExceptOne",
      "content": {
        "inboundId": "customer.app.variant1.InboundID"
      },
      "texts": {}
    }
  ]
}
"
`;

exports[`FLPConfigGenerator Integration Tests should generate FLP configuration successfully - VS Code 2`] = `
"#Make sure you provide a unique prefix to the newly added keys in this file, to avoid overriding of SAP Fiori application keys.

#XTIT: Application name
customer.app.variant1_sap.app.title=App Variant Title"
`;
